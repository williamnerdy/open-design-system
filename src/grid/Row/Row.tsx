import React from 'react';
import classNames from 'classnames';

type Aligns = 'middle' | 'bottom' | 'top';

type Justifies = 'left' | 'center' | 'right';

type Reverse = true;

type Mixed = { align?: Aligns; justify?: Justifies; reverse?: Reverse };

type Responsives = Mixed | Aligns | Reverse | Justifies;

interface Props {
  /** Children */
  children: React.ReactNode;
  /** Align content */
  align?: Aligns;
  /** justify */
  justify?: Justifies;
  /** Reverse */
  reverse?: Reverse;
  /** Small size */
  sm?: Responsives;
  /** Medium size */
  md?: Responsives;
  /** Large size */
  lg?: Responsives;
}

const getResponsive = (name: string, val?: Responsives) => {
  const align = typeof val === 'object' && val.align ? val.align : val;
  const justify = typeof val === 'object' && val.justify ? val.justify : val;
  const reverse = typeof val === 'object' && val.reverse ? val.reverse : val;

  return {
    [`${name}-${align}`]: align && typeof align !== 'object',
    [`${name}-${justify}`]: justify && typeof align !== 'object',
    [`${name}-${reverse}`]: reverse && typeof reverse !== 'object'
  };
};

const Row = ({ children, justify, align, reverse, sm, md, lg }: Props) => {
  const classes = classNames('row', {
    [`align-${align}`]: align,
    [`justify-${justify}`]: justify,
    [`reverse-${reverse}`]: reverse,
    ...getResponsive('sm', sm),
    ...getResponsive('md', md),
    ...getResponsive('lg', lg),
    reverse
  });

  return <div className={classes}>{children}</div>;
};

export default Row;
