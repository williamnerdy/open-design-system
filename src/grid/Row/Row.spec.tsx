import React from 'react';
import { cleanup, render } from '@testing-library/react';
import Row from './Row';

describe('<Row />', () => {
  const contentRow = 'Row content';
  beforeEach(cleanup);

  it('should render without prop', () => {
    const { asFragment, getByText } = render(<Row>{contentRow}</Row>);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(contentRow).classList.contains('row')).toBe(true);
  });

  describe('Align row', () => {
    it('should render with align in left', () => {
      const { asFragment, getByText } = render(
        <Row align="middle">Row content</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('align-middle')).toBe(
        true
      );
    });

    it('should render with align in bottom', () => {
      const { asFragment, getByText } = render(
        <Row align="bottom">Row content</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('align-bottom')).toBe(
        true
      );
    });

    it('should render with align in left', () => {
      const { asFragment, getByText } = render(
        <Row align="top">Row content</Row>
      );

      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('align-top')).toBe(true);
    });
  });

  describe('Justify', () => {
    it('should render with justify left', () => {
      const { asFragment, getByText } = render(
        <Row justify="left">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('justify-left')).toBe(
        true
      );
    });

    it('should render with justify center', () => {
      const { asFragment, getByText } = render(
        <Row justify="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('justify-center')).toBe(
        true
      );
    });

    it('should render with justify right', () => {
      const { asFragment, getByText } = render(
        <Row justify="right">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('justify-right')).toBe(
        true
      );
    });
  });

  describe('Justify by size', () => {
    it('should render with justify center sm', () => {
      const { asFragment, getByText } = render(
        <Row sm="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('sm-center')).toBe(true);
    });

    it('should render with justify center md', () => {
      const { asFragment, getByText } = render(
        <Row md="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('md-center')).toBe(true);
    });

    it('should render with justify center lg', () => {
      const { asFragment, getByText } = render(
        <Row lg="middle">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('lg-middle')).toBe(true);
    });
  });

  describe('Reverse row', () => {
    it('should render with reverse', () => {
      const { asFragment, getByText } = render(
        <Row reverse={true}>Row content</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('reverse-true')).toBe(
        true
      );
    });
  });

  describe('justify, align and reverse', () => {
    it('should render with justify center, aling middle, sm and reverse', () => {
      const { asFragment, getByText } = render(
        <Row sm={{ justify: 'center', align: 'middle', reverse: true }}>
          {contentRow}
        </Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('sm-center')).toBe(true);
      expect(getByText(contentRow).classList.contains('sm-middle')).toBe(true);
      expect(getByText(contentRow).classList.contains('sm-true')).toBe(true);
    });

    it('should render with justify center, aling middle, md and reverse', () => {
      const { asFragment, getByText } = render(
        <Row md={{ justify: 'center', align: 'middle', reverse: true }}>
          {contentRow}
        </Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('md-center')).toBe(true);
      expect(getByText(contentRow).classList.contains('md-middle')).toBe(true);
      expect(getByText(contentRow).classList.contains('md-true')).toBe(true);
    });

    it('should render with justify center, aling middle, lg and reverse', () => {
      const { asFragment, getByText } = render(
        <Row lg={{ justify: 'center', align: 'middle', reverse: true }}>
          {contentRow}
        </Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('lg-center')).toBe(true);
      expect(getByText(contentRow).classList.contains('lg-middle')).toBe(true);
      expect(getByText(contentRow).classList.contains('lg-true')).toBe(true);
    });
  });

  describe('Individual responsiveness of sm', () => {
    it('should render with sm justify center', () => {
      const { asFragment, getByText } = render(
        <Row sm="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('sm-center')).toBe(true);
    });

    it('should render with sm aling middle', () => {
      const { asFragment, getByText } = render(
        <Row sm="middle">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('sm-middle')).toBe(true);
    });

    it('should render with sm reverse', () => {
      const { asFragment, getByText } = render(
        <Row sm={{ reverse: true }}>{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('sm-true')).toBe(true);
    });
  });

  describe('Individual responsiveness of md', () => {
    it('should render with md justify center', () => {
      const { asFragment, getByText } = render(
        <Row md="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('md-center')).toBe(true);
    });

    it('should render with md aling middle', () => {
      const { asFragment, getByText } = render(
        <Row md="middle">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('md-middle')).toBe(true);
    });

    it('should render with md reverse', () => {
      const { asFragment, getByText } = render(
        <Row md={{ reverse: true }}>{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('md-true')).toBe(true);
    });
  });

  describe('Individual responsiveness of lg', () => {
    it('should render with lg justify center', () => {
      const { asFragment, getByText } = render(
        <Row lg="center">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('lg-center')).toBe(true);
    });

    it('should render with lg aling middle', () => {
      const { asFragment, getByText } = render(
        <Row lg="middle">{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('lg-middle')).toBe(true);
    });

    it('should render with lg reverse', () => {
      const { asFragment, getByText } = render(
        <Row lg={{ reverse: true }}>{contentRow}</Row>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText(contentRow).classList.contains('lg-true')).toBe(true);
    });
  });
});
