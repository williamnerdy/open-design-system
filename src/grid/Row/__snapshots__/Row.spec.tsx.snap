// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> Align row should render with align in bottom 1`] = `
<DocumentFragment>
  <div
    class="row align-bottom"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Align row should render with align in left 1`] = `
<DocumentFragment>
  <div
    class="row align-middle"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Align row should render with align in left 2`] = `
<DocumentFragment>
  <div
    class="row align-top"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of lg should render with lg aling middle 1`] = `
<DocumentFragment>
  <div
    class="row lg-middle"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of lg should render with lg justify center 1`] = `
<DocumentFragment>
  <div
    class="row lg-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of lg should render with lg reverse 1`] = `
<DocumentFragment>
  <div
    class="row lg-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of md should render with md aling middle 1`] = `
<DocumentFragment>
  <div
    class="row md-middle"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of md should render with md justify center 1`] = `
<DocumentFragment>
  <div
    class="row md-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of md should render with md reverse 1`] = `
<DocumentFragment>
  <div
    class="row md-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of sm should render with sm aling middle 1`] = `
<DocumentFragment>
  <div
    class="row sm-middle"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of sm should render with sm justify center 1`] = `
<DocumentFragment>
  <div
    class="row sm-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Individual responsiveness of sm should render with sm reverse 1`] = `
<DocumentFragment>
  <div
    class="row sm-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify by size should render with justify center lg 1`] = `
<DocumentFragment>
  <div
    class="row lg-middle"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify by size should render with justify center md 1`] = `
<DocumentFragment>
  <div
    class="row md-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify by size should render with justify center sm 1`] = `
<DocumentFragment>
  <div
    class="row sm-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify should render with justify center 1`] = `
<DocumentFragment>
  <div
    class="row justify-center"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify should render with justify left 1`] = `
<DocumentFragment>
  <div
    class="row justify-left"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Justify should render with justify right 1`] = `
<DocumentFragment>
  <div
    class="row justify-right"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> Reverse row should render with reverse 1`] = `
<DocumentFragment>
  <div
    class="row reverse-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> justify, align and reverse should render with justify center, aling middle, lg and reverse 1`] = `
<DocumentFragment>
  <div
    class="row lg-middle lg-center lg-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> justify, align and reverse should render with justify center, aling middle, md and reverse 1`] = `
<DocumentFragment>
  <div
    class="row md-middle md-center md-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> justify, align and reverse should render with justify center, aling middle, sm and reverse 1`] = `
<DocumentFragment>
  <div
    class="row sm-middle sm-center sm-true"
  >
    Row content
  </div>
</DocumentFragment>
`;

exports[`<Row /> should render without prop 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    Row content
  </div>
</DocumentFragment>
`;
